agent/engine/docker_task_engine.go

```diff
@@ -290,33 +290,39 @@ func (engine *DockerTaskEngine) synchronizeContainerStatus(container *api.Docker
 		} else {
 			container.DockerID = describedContainer.ID
 			container.Container.SetKnownStatus(dockerStateToState(describedContainer.State))
+			container.Container.SetCreatedAt(describedContainer.Created)
+			container.Container.SetStartedAt(describedContainer.State.StartedAt)
+			container.Container.SetFinishedAt(describedContainer.State.FinishedAt)
 			// update mappings that need dockerid
 			engine.state.AddContainer(container, task)
 			engine.imageManager.RecordContainerReference(container.Container)
+			engine.state.SetLabels(container.DockerID, describedContainer.Config.Labels)
 		}
 		return
 	}
 
-	if container.DockerID != "" {
-		currentState, metadata := engine.client.DescribeContainer(container.DockerID)
-		if metadata.Error != nil {
-			currentState = api.ContainerStopped
-			if !container.Container.KnownTerminal() {
-				container.Container.ApplyingError = api.NewNamedError(&ContainerVanishedError{})
-				log.Warn("Could not describe previously known container; assuming dead", "err", metadata.Error, "id", container.DockerID, "name", container.DockerName)
-				engine.imageManager.RemoveContainerReferenceFromImageState(container.Container)
-			}
-		} else {
-			engine.imageManager.RecordContainerReference(container.Container)
-			if engine.cfg.ContainerMetadataEnabled && !container.Container.IsMetadataFileUpdated() {
-				go engine.updateMetadataFile(task, container)
-			}
+	currentState, metadata := engine.client.DescribeContainer(container.DockerID)
+	if metadata.Error != nil {
+		currentState = api.ContainerStopped
+		if !container.Container.KnownTerminal() {
+			container.Container.ApplyingError = api.NewNamedError(&ContainerVanishedError{})
+			log.Warn("Could not describe previously known container; assuming dead", "err", metadata.Error, "id", container.DockerID, "name", container.DockerName)
+			engine.imageManager.RemoveContainerReferenceFromImageState(container.Container)
 		}
-		if currentState > container.Container.GetKnownStatus() {
-			// update the container known status
-			container.Container.SetKnownStatus(currentState)
+	} else {
+		engine.state.SetLabels(metadata.DockerID, metadata.Labels)
+		container.Container.SetCreatedAt(metadata.CreatedAt)
```

Comment:
nit suggestion: Should we have a single method for setting all the timestamps, since they use the same lock?

Reply:
I initially had it that way, but separated them out as I can see use-cases where it'd be required to override only one of these.